File: scripts/report.py
"""Escaped, self-contained HTML reports from explicitly selected local CRM data."""
import datetime as dt
from decimal import Decimal, localcontext
from html import escape
import json
import re
import urllib.parse


def cost(amount, currency):
    if amount is None and currency is None:
        return None, None
    if not isinstance(amount, str) or not re.fullmatch(r"(?:0|[1-9][0-9]{0,17})(?:\.[0-9]{1,6})?", amount):
        raise ValueError("cost_amount must be a nonnegative decimal string with at most six decimal places")
    if not isinstance(currency, str) or not re.fullmatch(r"[A-Z]{3}", currency):
        raise ValueError("cost_currency must be an uppercase three-letter currency label paired with cost_amount")
    return decimal_text(Decimal(amount)), currency


def decimal_text(value):
    text = format(value, "f")
    return text.rstrip("0").rstrip(".") if "." in text else text


def escaped(value):
    return escape(str(value) if value is not None else "", quote=True)


def link(value):
    try:
        parsed = urllib.parse.urlsplit(value)
        if parsed.scheme not in {"https", "http"} or not parsed.hostname or parsed.username or parsed.password or any(ord(char) < 32 for char in value):
            raise ValueError()
    except (ValueError, TypeError):
        return escaped(value)
    return '<a href="' + escaped(value) + '" target="_blank" rel="noopener noreferrer">' + escaped(value) + '</a>'


def render(connection, title, brand, generated_at, project_id=None, include_notes=False, include_costs=False):
    for name, value in (("title", title), ("brand", brand)):
        if not isinstance(value, str) or not value.strip() or len(value) > 300 or any(ord(char) < 32 for char in value):
            raise ValueError(name + " must be a nonempty label of at most 300 characters")
    query = "SELECT * FROM placements" + (" WHERE project_id=?" if project_id else "") + " ORDER BY COALESCE(project_id,''),source_url,target_url,id"
    connection.execute("BEGIN")
    try:
        placements = [dict(row) for row in connection.execute(query, (project_id,) if project_id else ())]
        connection.execute("COMMIT")
    except Exception:
        connection.execute("ROLLBACK")
        raise
    counts, totals, rows, observed_dates = {}, {}, [], []
    for placement in placements:
        cloud = json.loads(placement["cloud_state"])
        if not isinstance(cloud, dict) or not isinstance(cloud.get("state", "unobserved"), str):
            raise ValueError("A placement has malformed cloud state; repair the CRM before reporting")
        state = cloud.get("state", "unobserved")
        counts[state] = counts.get(state, 0) + 1
        observed_at = placement["cloud_observed_at"]
        if observed_at:
            observed_dates.append(observed_at)
        columns = [link(placement["source_url"]), link(placement["target_url"]), escaped(str(state).replace("_", " ")), escaped(observed_at or "No observation")]
        if include_notes:
            columns.append(escaped(placement["notes"]))
        if include_costs:
            amount, currency = cost(placement["cost_amount"], placement["cost_currency"])
            if amount is not None:
                with localcontext() as context:
                    context.prec = 50
                    totals[currency] = totals.get(currency, Decimal(0)) + Decimal(amount)
                columns.append(escaped(currency + " " + amount))
            else:
                columns.append("Not recorded")
        rows.append("<tr>" + "".join("<td>" + column + "</td>" for column in columns) + "</tr>")
    headings = ["Source page", "Destination", "Observed state", "Observation time (UTC offset shown)"]
    if include_notes:
        headings.append("Local notes (included by request)")
    if include_costs:
        headings.append("Recorded local cost")
    if not rows:
        rows.append('<tr><td colspan="' + str(len(headings)) + '">No placements match this report selection.</td></tr>')
    privacy = "Includes source and destination URLs, observed states and observation times."
    privacy += " Local notes are included by request." if include_notes else " Local notes are excluded."
    privacy += " Recorded costs are included by request." if include_costs else " Recorded costs are excluded."
    scope = "Selected project" if project_id else "All local placements"
    coverage = "No cloud observations are available for these placements."
    if observed_dates:
        chronological = sorted(observed_dates, key=lambda value: dt.datetime.fromisoformat(value.replace("Z", "+00:00")))
        coverage = "Observed timestamps range from " + chronological[0] + " to " + chronological[-1] + "."
    totals_html = ""
    if include_costs:
        totals_html = '<section class="costs"><h2>Recorded costs</h2>'
        if totals:
            totals_html += '<ul>' + ''.join('<li>' + escaped(currency + " " + decimal_text(amount)) + '</li>' for currency, amount in sorted(totals.items())) + '</ul>'
        else:
            totals_html += '<p>No costs are recorded for the selected placements.</p>'
        totals_html += '<p>Totals combine recorded amounts within each currency. No currency conversion, market valuation or tax calculation is applied.</p></section>'
    document = '''<!doctype html>
<html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width,initial-scale=1">
<meta name="referrer" content="no-referrer"><meta name="robots" content="noindex,nofollow">
<meta http-equiv="Content-Security-Policy" content="default-src 'none'; style-src 'unsafe-inline'; base-uri 'none'; form-action 'none'">
<title>''' + escaped(title) + '''</title>
<style>
*{box-sizing:border-box}body{margin:0;color:#192b36;background:#f7f9fb;font:14px/1.65 -apple-system,BlinkMacSystemFont,"Segoe UI",sans-serif}main{max-width:1400px;margin:auto;padding:48px 36px}.brand{font:12px/1.5 ui-monospace,SFMono-Regular,Consolas,monospace;letter-spacing:.06em;text-transform:uppercase;color:#c4472d}h1,h2{font-weight:600;line-height:1.12;letter-spacing:-.04em}h1{font-size:44px;margin:18px 0 20px}h2{font-size:27px}.meta,.coverage,.disclosure{color:#566570;font-size:13px;max-width:1000px}.coverage{border-left:2px solid #c4472d;padding-left:16px;margin-top:24px}.metrics{display:flex;flex-wrap:wrap;margin:32px 0;border-block:1px solid #d8dfe5}.metric{flex:1;min-width:160px;padding:20px 24px 20px 0}.metric+.metric{border-left:1px solid #d8dfe5;padding-left:24px}.metric strong{display:block;font-size:32px;font-weight:550;letter-spacing:-.03em}.metric span{font-size:12px;color:#566570}.table-region{overflow-x:auto;max-width:100%}table{border-collapse:collapse;width:100%;min-width:760px;background:#fff;font-size:13px;table-layout:fixed}caption{text-align:left;font-size:19px;font-weight:600;letter-spacing:-.025em;padding:0 0 16px}th,td{text-align:left;border-bottom:1px solid #d8dfe5;padding:15px;vertical-align:top;overflow-wrap:anywhere;white-space:pre-wrap}th{background:#eef2f5;color:#566570;font:11px/1.6 ui-monospace,SFMono-Regular,Consolas,monospace}a{color:#192b36;text-underline-offset:3px}a:focus-visible{outline:3px solid #c4472d;outline-offset:3px}footer{border-top:1px solid #d8dfe5;margin-top:36px;padding-top:20px;font-size:12px;color:#566570}.costs{margin-top:32px}.costs p{font-size:13px;color:#566570}ul{padding-left:20px}@media(max-width:700px){main{padding:32px 20px}h1{font-size:34px}.metric{min-width:50%;padding:16px 12px}.metric+.metric{padding-left:12px}.metric:nth-child(odd){border-left:0}.metric:nth-child(n+3){border-top:1px solid #d8dfe5}}@media print{body{background:white}main{padding:0;max-width:none}.metric,th{background:#eef2f5}.table-region{overflow:visible}table{min-width:0;font-size:10px}th,td{padding:9px}th{font-size:9px}tr{break-inside:avoid}thead{display:table-header-group}a{color:#192b36;text-decoration:none}@page{size:landscape;margin:12mm}}
</style></head><body><main>
<div class="brand">''' + escaped(brand) + '''</div><h1>''' + escaped(title) + '''</h1>
<p class="meta">Generated ''' + escaped(generated_at) + " · " + escaped(scope) + ''' · Local CRM snapshot</p>
<p class="coverage">''' + escaped(coverage) + ''' This report does not perform new checks. Unknown or unobserved results do not establish link loss.</p>
<div class="metrics"><div class="metric"><strong>''' + str(len(placements)) + '''</strong><span>Selected placements</span></div><div class="metric"><strong>''' + str(counts.get("present", 0)) + '''</strong><span>Observed present</span></div><div class="metric"><strong>''' + str(counts.get("confirmed_missing", 0)) + '''</strong><span>Confirmed missing</span></div><div class="metric"><strong>''' + str(counts.get("unknown", 0) + counts.get("unobserved", 0)) + '''</strong><span>Unknown or unobserved</span></div></div>
<div class="table-region" role="region" aria-label="Selected placement observations" tabindex="0"><table><caption>Selected placement observations</caption><thead><tr>''' + ''.join('<th scope="col">' + escaped(heading) + '</th>' for heading in headings) + '''</tr></thead><tbody>''' + ''.join(rows) + '''</tbody></table></div>
''' + totals_html + '''<footer><p>''' + escaped(privacy) + ''' Contacts, messages, API credentials, disavow rules and raw cloud event payloads are not included.</p><p>Created with the local AgentLinkOps toolkit. This is an HTML report; browser print options are separate from report generation.</p></footer>
</main></body></html>
'''
    return document, {"format": "html", "placements": len(placements), "project_id": project_id, "generated_at": generated_at, "included_local_notes": include_notes, "included_local_costs": include_costs}

File: cli/report.js
// `agentlinkops report`: one self-contained HTML file, reproducible from a frozen dataset.
//
// **Reproducible is the requirement, and it is the one that constrains the design.** The same
// ledger and the same observations must produce byte-identical output, because a report a
// customer sends to a client is a document they may have to produce again months later and
// defend as unchanged. So: no clock read that is not an input, no random ids, stable ordering,
// and a content digest printed in the footer so "identical" is checkable rather than claimed.
//
// **A report timestamp does not make its observations fresh.** Every row carries the date IT was
// observed, and the header states the range, because the single most misleading thing a link
// report can do is print today's date above a figure from March.
//
// Self-contained: no external stylesheet, font, script or image. That makes it openable offline
// forever, printable to PDF from any browser, and incapable of telling anyone it was opened.
//
// v2 makes the document a RECEIPT (DP-0004-T20): every row carries the observation behind it —
// when it was observed, the content hash of what was captured, who checked, and a stable
// evidence reference that resolves to the observation itself. One dataset feeds both exports:
// this HTML view for a human client, and `report --json` for an agent, so the two can never
// disagree about what was observed.
import {indexRowsForUrl} from './index-observations.js';
import {reconcileIndexObservations} from '../src/index-observation.js';
import { createHash } from 'node:crypto';
import { latestByEntry } from './mirror.js';
import { mirrorEvidenceReference } from '../src/evidence-reference.js';

export const REPORT_VERSION = 2;

/** HTML-escape. Every value below is publisher or customer content and none of it is trusted. */
export const escape = value => String(value ?? '')
  .replaceAll('&', '&amp;').replaceAll('<', '&lt;').replaceAll('>', '&gt;')
  .replaceAll('"', '&quot;').replaceAll("'", '&#39;');

const STATE_LABEL = Object.freeze({
  present: 'Present', absent: 'Not found', unknown: 'Could not check',
  source_unavailable: 'Page unavailable', unchecked: 'Never checked',
});

/**
 * The frozen dataset a report renders.
 *
 * Selection happens here and nowhere else, so a caller can hold this object, assert it, and know
 * the rendered document is a pure function of it.
 */
export function freezeDataset(entries, observations, { asOf, title = 'Backlink report', includeNotes = false, includeRetired = false, indexObservations = [] } = {}) {
  const latest = latestByEntry(observations);
  const rows = entries
    .filter(entry => includeRetired || entry.intent !== 'retired')
    .map(entry => {
      const row = latest.get(entry.id) ?? null;
      const indexRows=indexRowsForUrl(indexObservations,entry.source);
      const origin = row?.source === 'cloud' ? 'cloud' : (row?.source ?? null);
      return {
        id: entry.id, intent: entry.intent, source: entry.source, target: entry.target, scope: entry.scope,
        expected_anchor: entry.expect?.anchor ?? null,
        state: row?.state ?? 'unchecked',
        reason: row?.reason ?? null,
        occurrences: row?.occurrences ?? 0,
        checked_at: row?.checked_at ?? null,
        checked_by: row?.source ?? null,
        anchor: row?.result?.occurrences?.[0]?.anchor ?? null,
        rel: row?.result?.occurrences?.[0]?.rel ?? [],
        // The receipt block: what was observed, when, by which check, and how to point at the
        // observation itself. Null on a never-checked row — an absent receipt is a fact here,
        // not a missing field.
        evidence: row ? {
          observed_at: row.checked_at ?? null,
          // The content hash of the captured document, present whenever the check recorded one.
          // This is the field anchoring (DP-0042-T07) will make independently verifiable; it is
          // in the export now so a document made today can be verified the day anchoring lands.
          content_hash: row.result?.evidence?.sha256 ?? null,
          // The hosted snapshot key, when this observation was synced from the workspace.
          // Local checks never store publisher bytes anywhere, so null there means exactly that.
          snapshot_key: row.evidence_key ?? null,
          // Who ran the check: a local CLI run or the hosted scheduler. Two origins at the same
          // instant are two observations, and a receipt that blurred them would blur method.
          origin,
          checker_version: row.checker_version ?? row.result?.evidence?.checkerVersion ?? null,
          // Named, not implied: a static observation is not a claim about what a reader sees.
          method: row.result?.evidence?.rendered === true || row.result?.evidence?.method === 'browser_html'
            ? 'rendered' : row.result?.evidence?.rendered === false || row.result?.evidence?.method === 'http_html'
              ? 'static_html' : null,
          // The hosted observation id when the mirror row carries one (cloud-synced rows).
          hosted_observation_id: row.cloud_observation_id ?? null,
          // The stable reference that resolves to this observation from the repository mirror.
          locator: mirrorEvidenceReference({ entryId: entry.id, checkedAt: row.checked_at ?? undefined, origin: origin ?? undefined }),
        } : null,
        // Opt-in, because a note can hold commercial context — what a placement cost, what a
        // publisher said — and a report is the document most likely to be forwarded.
        ...(indexRows.length?{index_observations:indexRows,index_reconciliation:reconcileIndexObservations(indexRows)}:{}),
        note: includeNotes ? (entry.note ?? null) : null,
        ref: includeNotes ? (entry.ref ?? null) : null,
      };
    })
    // Stable ordering, by state then id: byte-identical output needs a total order with no ties
    // broken by insertion.
    .sort((a, b) => (a.state === b.state ? (a.id < b.id ? -1 : 1) : (a.state < b.state ? -1 : 1)));

  const dates = rows.map(row => row.checked_at).filter(Boolean).sort();
  const counts = rows.reduce((acc, row) => { acc[row.state] = (acc[row.state] ?? 0) + 1; return acc; }, {});
  return {
    v: REPORT_VERSION, title, as_of: asOf, include_notes: includeNotes,
    rows,
    totals: { entries: rows.length, ...counts },
    // Stated rather than implied. A report over a ledger nobody has checked is not a report about
    // links, and the reader should see that before the table.
    coverage: {
      checked: rows.filter(row => row.checked_at).length,
      never_checked: rows.filter(row => !row.checked_at).length,
      observed_from: dates[0] ?? null,
      observed_to: dates.at(-1) ?? null,
    },
  };
}

const STYLE = `
:root{color-scheme:light}
*{box-sizing:border-box}
body{margin:0;padding:32px;font:14px/1.5 -apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Helvetica,Arial,sans-serif;color:#1a1a1a;background:#fff}
h1{font-size:22px;margin:0 0 4px}
.sub{color:#555;margin:0 0 24px}
.cards{display:flex;flex-wrap:wrap;gap:12px;margin:0 0 24px}
.card{border:1px solid #ddd;border-radius:6px;padding:10px 14px;min-width:120px}
.card b{display:block;font-size:20px}
.card span{color:#555;font-size:12px}
table{border-collapse:collapse;width:100%;font-size:13px}
th,td{text-align:left;vertical-align:top;padding:8px 10px;border-bottom:1px solid #eee}
th{background:#fafafa;border-bottom:1px solid #ddd;font-weight:600}
td.url{word-break:break-all;max-width:340px}
.state{display:inline-block;padding:1px 8px;border-radius:10px;font-size:12px;border:1px solid #ccc}
.present{background:#eef7ee;border-color:#b6d7b6}
.absent{background:#fdeeee;border-color:#e2b6b6}
.unknown,.unchecked,.source_unavailable{background:#f6f6f6}
code.hash{font:12px/1.4 ui-monospace,SFMono-Regular,Menlo,Consolas,monospace}
.ref{font:11px/1.4 ui-monospace,SFMono-Regular,Menlo,Consolas,monospace;word-break:break-all}
.note{color:#555;font-size:12px}
footer{margin-top:28px;padding-top:12px;border-top:1px solid #eee;color:#555;font-size:12px}
.caveat{margin:18px 0;padding:10px 14px;border-left:3px solid #ddd;color:#444;font-size:13px}
@media print{body{padding:0}th{background:#fff}tr{break-inside:avoid}.card{break-inside:avoid}}
`.trim();

/**
 * Renders the frozen dataset. A pure function of its argument: same input, same bytes.
 *
 * The digest in the footer is over the DATASET rather than over the HTML, so it stays stable
 * across a cosmetic change to this template while still changing the moment a figure does.
 */
export function renderReport(dataset, { brand = null } = {}) {
  const digest = createHash('sha256').update(JSON.stringify(dataset)).digest('hex');
  const card = (label, value) => `<div class="card"><b>${escape(value)}</b><span>${escape(label)}</span></div>`;
  const coverage = dataset.coverage;
  // The receipt column. "hash not recorded" is stated rather than papered over: a local check
  // that never completed a fetch has no hash, and a cloud-synced row may carry its snapshot key
  // without a hash in an older event. Neither is a claim about the link.
  const evidenceCell = row => {
    if (!row.evidence) return '<span class="note">no observation</span>';
    const hash = row.evidence.content_hash ? `<code class="hash">${escape(row.evidence.content_hash.slice(0, 12))}</code>` : '<span class="note">hash not recorded</span>';
    const origin = row.evidence.origin === 'cloud' ? 'hosted check' : row.evidence.origin === 'local' ? 'local check' : 'unknown origin';
    const checker = row.evidence.checker_version ? ` &middot; checker ${escape(row.evidence.checker_version)}` : '';
    const method=row.evidence.method==='rendered'?'browser-rendered HTML':row.evidence.method==='static_html'?'static HTML':'method not recorded';
    return `${hash}<div class="note">${escape(origin)}${checker}</div><div class="note">${method}</div>`;
  };
  const hasIndex=dataset.rows.some(row=>row.index_observations?.length);
  const rows = dataset.rows.map(row => `<tr>
      <td class="url"><a href="${escape(row.source)}">${escape(row.source)}</a></td>
      <td class="url">${escape(row.target)}</td>
      <td><span class="state ${escape(row.state)}">${escape(STATE_LABEL[row.state] ?? row.state)}</span></td>
      <td>${row.anchor ? escape(row.anchor) : '<span class="note">&mdash;</span>'}${(row.rel ?? []).length ? `<div class="note">rel: ${escape(row.rel.join(' '))}</div>` : ''}</td>
      <td>${row.checked_at ? escape(row.checked_at.slice(0, 10)) : '<span class="note">never</span>'}</td>
      <td>${evidenceCell(row)}</td>
      ${hasIndex?`<td>${(row.index_observations??[]).map(o=>`${escape(o.tier)} · ${escape(o.reason)}<div>${escape(o.checked_at)} · ${escape(o.backend)}</div><div class="ref">${escape(o.locator)}</div>`).join('<hr>')||'Not checked'}</td>`:''}
      ${dataset.include_notes ? `<td class="note">${escape(row.ref ?? '')}${row.note ? `<div>${escape(row.note)}</div>` : ''}</td>` : ''}
    </tr>`).join('\n');
  // Every row's full reference, printable: the table above stays readable while the document a
  // client forwards still carries the complete handle for each observation behind it.
  const receiptRows = dataset.rows.filter(row => row.evidence?.locator).map(row => `<tr>
      <td>${escape(row.id)}</td>
      <td>${escape(row.evidence.observed_at ?? 'unknown')}</td>
      <td>${row.evidence.content_hash ? `<code class="hash">${escape(row.evidence.content_hash)}</code>` : '<span class="note">not recorded</span>'}</td>
      <td class="ref">${escape(row.evidence.locator)}</td>
    </tr>`).join('\n');

  return `<!doctype html>
<html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width,initial-scale=1">
<title>${escape(dataset.title)}</title><style>${STYLE}</style></head>
<body>
<h1>${escape(dataset.title)}</h1>
<p class="sub">${brand ? `${escape(brand)} &middot; ` : ''}Prepared ${escape(dataset.as_of)}</p>
<div class="cards">
${card('links in report', dataset.totals.entries)}
${card('present', dataset.totals.present ?? 0)}
${card('not found', dataset.totals.absent ?? 0)}
${card('could not check', (dataset.totals.unknown ?? 0) + (dataset.totals.source_unavailable ?? 0))}
${card('never checked', dataset.totals.unchecked ?? 0)}
</div>
<p class="caveat"><strong>What these dates mean.</strong> Every row shows the date that link was
last observed, between ${escape(coverage.observed_from ?? 'n/a')} and ${escape(coverage.observed_to ?? 'n/a')}.
The date at the top is when this document was prepared and does not make any observation newer
than it is. ${coverage.never_checked ? `${coverage.never_checked} of ${dataset.totals.entries} links have never been checked and are listed as such.` : ''}</p>
<p class="caveat"><strong>What a result means.</strong> &ldquo;Present&rdquo; means the link was in
the observed HTML. &ldquo;Could not check&rdquo; is not the same as
&ldquo;not found&rdquo;: it means the page could not be read, and no conclusion was drawn.
Each receipt identifies the check method when it was recorded. A snapshot alone does not prove the link was visible on screen.</p>
<table><thead><tr>
<th>Source page</th><th>Target</th><th>Result</th><th>Anchor</th><th>Observed</th><th>Evidence</th>${hasIndex?'<th>Index evidence</th>':''}${dataset.include_notes ? '<th>Notes</th>' : ''}
</tr></thead><tbody>
${rows}
</tbody></table>
${receiptRows ? `<h2>Evidence references</h2>
<p class="caveat"><strong>What a reference is.</strong> Each row above was concluded from one
saved observation. Its reference names that observation — which ledger entry, when it was
checked, and whether the check ran locally or on the hosted service — and resolves to the
recorded result from the repository that produced this document, with no account required. The
content hash, where recorded, is the hash of the document that was captured at that moment;
two captures of a changed page produce different hashes, which is what makes "the same
evidence" checkable rather than claimed.</p>
<table><thead><tr><th>Link</th><th>Observed at</th><th>Content hash</th><th>Reference</th></tr></thead><tbody>
${receiptRows}
</tbody></table>` : ''}
<footer>
AgentLinkOps report v${dataset.v}. Dataset digest <code>${escape(digest)}</code> &mdash; the same
ledger and observations always produce this digest. The same dataset exports as JSON
(<code>agentlinkops report --json</code>) and carries the same digest.
${dataset.include_notes ? 'Private notes are included in this document.' : 'Private notes and references are excluded.'}
</footer>
</body></html>
`;
}

/** The digest a caller can compare without rendering. */
export const datasetDigest = dataset => createHash('sha256').update(JSON.stringify(dataset)).digest('hex');
